### DISCUSSION TITLE
CMYK ↔ CIELAB conversions

### DISCUSSION THREAD

**Discussion by Kallys** [ImageMagick/ImageMagick#4412] [Help] [answered]
Dear ImageMagick community,

I'm working with CMYK and CIELAB colorspaces, and I'm searching for advices to treat them properly (ie, perceptual colors fidelity). Since I'm familiar to Photoshop, my first feeling is to reproduce its processing. So here are my experimentations:

# From CMYK to Lab conversion
| # | Generation process |  <img width="200">L |  <img width="200">a | <img width="200">b |
| - | - | - | - | - |
| 1 | 1. Creating 16bits "cmyk(100%,0%,0%,0%)" using Photoshop<br>2. Converting colorspace to Lab<br>3. Saving file to "cyan2lab-1.tiff" | 57.3327 | -40.1907 | -51.3735 |
| 2 | 1. ```$ magick "xc:cmyk(100%,0%,0%,0%)" -profile Coated\ FOGRA39\ \(ISO\ 12647-2-2004\).icc cyan.tiff```<br> 2. Opening "cyan.tiff" with Photoshop <br> 3. Converting colorspace to Lab<br> 4. Saving file to "cyan2lab-2.tiff" | 57.3327 | -40.1907 | -51.3735 |
| 3 | ```$ magick "xc:cmyk(100%,0%,0%,0%)" -profile Coated\ FOGRA39\ \(ISO\ 12647-2-2004\).icc cyan.tiff && tificc -v -e -w 16 -t 1 -b -c 2 -i Coated\ FOGRA39\ \(ISO\ 12647-2-2004\).icc -o*Lab cyan.tiff cyan2lab.tiff && magick cyan2lab.tiff -set colorspace lab txt:``` | 57.3297 | -39.858 | -51.0837 |
| 4 | ```$ magick "xc:cmyk(100%,0%,0%,0%)" -profile Coated\ FOGRA39\ \(ISO\ 12647-2-2004\).icc cyan.tiff && tificc -v -e -w 16 -t 1 -b -c 2 -i Coated\ FOGRA39\ \(ISO\ 12647-2-2004\).icc -o*Lab cyan.tiff cyan2lab.tiff && magick cyan2lab.tiff -set colorspace lab cyan2lab-4.tiff && magick cyan2lab-4.tiff txt:``` | 57.3297 | -39.856 | -51.0817 |
| 5 | ```$ magick "xc:cmyk(100%,0%,0%,0%)" -profile Coated\ FOGRA39\ \(ISO\ 12647-2-2004\).icc -profile GenericLabProfile.icc txt:``` | 57.3449 | -167.823 |-179.064 |
| 6 | ```$ magick "xc:cmyk(100%,0%,0%,0%)" -profile Coated\ FOGRA39\ \(ISO\ 12647-2-2004\).icc -profile GenericLabProfile.icc cyan2lab-6.tiff && magick cyan2lab-6.tiff txt:``` | 57.3449 | 87.179 |75.9377 |
| 7 | ```$ magick "xc:cmyk(100%,0%,0%,0%)" -profile Coated\ FOGRA39\ \(ISO\ 12647-2-2004\).icc -colorspace lab txt:``` | 91.1132 | -48.0875 | -14.1312 |
| 8 | ```$ magick "xc:cmyk(100%,0%,0%,0%)" -profile Coated\ FOGRA39\ \(ISO\ 12647-2-2004\).icc -colorspace lab cyan2lab-8.tiff && magick cyan2lab-8.tiff txt:``` | 91.1132 | -48.0856 | -14.1323 |
| 9 | ```$ magick "xc:cmyk(100%,0%,0%,0%)" -black-point-compensation -intent relative -profile Coated\ FOGRA39\ \(ISO\ 12647-2-2004\).icc -profile Lab-D50-Identity-elle-V4.icc -channel GB -evaluate Add %[fx:\(QuantumRange+1\)/2] +channel cyan2lab-9.tiff && magick cyan2lab-9.tiff txt:``` | 57.3297 | -40.358 | -51.5837 |
| 10 | ```$ magick "xc:cmyk(100%,0%,0%,0%)" -profile Coated\ FOGRA39\ \(ISO\ 12647-2-2004\).icc -profile Lab-D50-Identity-elle-V4.icc -channel GB -evaluate Add %[fx:\(QuantumRange+1\)/2] +channel cyan2lab-10.tiff && magick cyan2lab-10.tiff txt:``` | 57.3449 | -40.323 | -51.5603 |

**Note:** L, a and b components are always displayed using ImageMagick to `txt:` conversion.

### Comments
* Writing file seems to have a little influence on components
* I'm unable to match exactly Photoshop behaviour
* I'm able to get close to Photoshop behaviour using `tificc`
* I'm **unable to get close to Photoshop behaviour using only ImageMagick**

# From Lab to CMYK conversion
| # | Generation process | <img width="200">C | <img width="200">M | <img width="200">Y | <img width="200">K |
| - | - | - | - | - | - |
| 1 | 1. Creating 16bits "cmyk(100%,0%,0%,0%)" using Photoshop<br>2. Converting colorspace to Lab<br>3. Saving file to tiff<br>4. Reopening file<br>5. Converting colorspace to CMYK<br> 6. Saving file to "cyan2lab2cmyk-1.tiff" | 254.953 | 5.65759 | 3.60311 | 0.0466926 |
| 2 | 1. Opening "cyan2lab-4.tiff" with Photoshop<br>2. Converting colorspace to Lab<br>3. Saving file to "cyan2lab2cmyk-2.tiff" | 254.097 | 6.5214 | 3.76654 | 0.0466926 |
| 3 | ```$ tificc -e -w 16 -t 1 -n -i*Lab -b -c 2 -v -o Coated\ FOGRA39\ \(ISO\ 12647-2-2004\).icc cyan2lab-4.tiff cyan2lab2cmyk-3.tiff && magick cyan2lab2cmyk-3.tiff txt:``` | 249.156 | 8.89105 | 4.57588 | 0.155642 |
| 4 | ```$ magick cyan2lab-4.tiff -intent relative -profile GenericLabProfile.icc -profile Coated\ FOGRA39\ \(ISO\ 12647-2-2004\).icc cyan2lab2cmyk-4.tiff && magick cyan2lab2cmyk-4.tiff txt:``` | 255 | 9.15953 | 5.59533 | 0.206226 |
| 5 | ```$ magick cyan2lab-4.tiff -profile Coated\ FOGRA39\ \(ISO\ 12647-2-2004\).icc -colorspace cmyk txt:``` | 1178.07 | 74.9525 | 0 | 28.7747 |
| 6 | ```$ magick cyan2lab-4.tiff -colorspace sRGB -intent relative -profile sRGB\ IEC61966-21.icc -profile Coated\ FOGRA39\ \(ISO\ 12647-2-2004\).icc txt:``` | 194.969 | 54.9416 | 0 | 0 |
| 7 | ```$ magick cyan2lab-4.tiff -black-point-compensation -intent relative -profile Lab-D50-Identity-elle-V4.icc -profile Coated\ FOGRA39\ \(ISO\ 12647-2-2004\).icc cyan2lab2cmyk-7.tiff && magick cyan2lab2cmyk-7.tiff txt:``` | 251.911 | 8.20623 | 4.18677 | 0.0350195 |
| 8 | ```$ magick cyan2lab-4.tiff -profile Lab-D50-Identity-elle-V4.icc -profile Coated\ FOGRA39\ \(ISO\ 12647-2-2004\).icc cyan2lab2cmyk-8.tiff && magick cyan2lab2cmyk-8.tiff txt:``` | 250.198 | 34.716 | 20.5058 | 0.645914 |

**Note:** C, M, Y and K components are always displayed using ImageMagick to `txt:` conversion.

### Comments
* Writing file does not seem to have an influence on components
* I'm unable to get back to full cyan CMYK image
* I'm unable to properly convert to Lab colorspace without using a fake "GenericLabProfile.icc" found somewhere<sup>1</sup> on Internet
* I'm unable to get really close to Photoshop behaviour
* This time, full ImageMagick method is quite satisfying

# Notes
* Photoshop color settings
![image](https://user-images.githubusercontent.com/1195845/138841223-92eb41ce-a1e5-481f-a6e1-bef818403cfb.png)
* Photoshop profile conversion to CMYK (Fogra39) settings (same settings except profile when converting to Lab)
![image](https://user-images.githubusercontent.com/1195845/138872236-a4e92a51-84d6-46cc-b80b-20ad7eead765.png)
* ImageMagick version:
```
Version: ImageMagick 7.1.0-10 Q16-HDRI x86_64 2021-10-05 https://imagemagick.org
Copyright: (C) 1999-2021 ImageMagick Studio LLC
License: https://imagemagick.org/script/license.php
Features: Cipher DPC HDRI OpenMP(4.5)
Delegates (built-in): bzlib djvu freetype jbig jng jpeg lcms lzma png tiff xml zlib
Compiler: gcc (10.2)
```
* [profiles.zip](https://github.com/ImageMagick/ImageMagick/files/7418059/profiles.zip) contains used profiles
* [somewhere<sup>1</sup>](https://legacy.imagemagick.org/discourse-server/viewtopic.php?f=1&t=25383&start=15)

# Questions
1. Is there any recommended method?
1. Is there "more accurate" methods than I exposed here?
1. Any comment or suggestion to get closer to Photoshop's behaviour?

Thanks!

**Comment by snibgo** [comment]
In your first table, you show numbers on the right. The columns are L* and a* and b*. But what are the four rows in each group? In #6, you have 4 L* values and 3 each of a* and b*. Why?

What is the nominal scale of the values? Are they 0 to 255, or 0 to 100, or what? Are IM and PS numbers using the same nominal scale?

Similar question apply to the second table, of CMYK.

IM records a* and b* numbers in memory in two different ways.

1. If "-colorspace Lab" has been used, the nominal range is 0 to 100% of QuantumRange.

2. If "-profile" has been used, the nominal range is -50% to +50% of QuantumRange.

When writing values to txt, IM always subtracts 50% of the values before showing them. For (1), this is reasonable, resulting in a nominal range of -50% to +50% of QuantumRange. For (2), this is confusing, resulting in a nominal range of -100% to 0% of QuantumRange.

**Reply by Kallys** [reply]
Thanks for your quick reply!

> In your first table, you show numbers on the right. The columns are L* and a* and b*. But what are the four rows in each group? In #6, you have 4 L* values and 3 each of a* and b*. Why?

I'm not sure to understand, each cell contains only one component as floating value (maybe that is a github rendering table issue that is confusing so I updated my first post and tried to force column width).

> What is the nominal scale of the values? Are they 0 to 255, or 0 to 100, or what? Are IM and PS numbers using the same nominal scale?

I always used IM to display components using same method: `magick <image> txt:`.  
I guess ranges are [0; 100] for _L_ component, and [-128; 127] for _a_ and _b_ components.

> Similar question apply to the second table, of CMYK.

C, M, Y and K components range appears to be in [0; 255].

> IM records a* and b* numbers in memory in two different ways.
>    1. If "-colorspace Lab" has been used, the nominal range is 0 to 100% of QuantumRange.
>    2. If "-profile" has been used, the nominal range is -50% to +50% of QuantumRange.
>
> When writing values to txt, IM always subtracts 50% of the values before showing them. For (1), this is reasonable, resulting in a nominal  range of -50% to +50% of QuantumRange. For (2), this is confusing, resulting in a nominal range of -100% to 0% of QuantumRange.

Ok, thanks for this clarification.

**Comment by snibgo** [comment]
Your Lab test # 7 is:
```
$ magick "xc:cmyk(100%,0%,0%,0%)" -profile Coated\ FOGRA39\ \(ISO\ 12647-2-2004\).icc -colorspace lab txt:
```
Test # 8 is similar. You are mixing two methods: "-profile" and "-colorspace". Don't do this. The methods are not interchangeable. Unless you take great care, you should not use both methods in the same command.

Tests # 1 to # 4 give similar results. The values seem to be on a scale of -50 to +50. The small differences may come from variations in processing the ICC metadata, or black point compensation, or intents.

Test # 5 has a* and b* values 50% lower than in # 1 to # 4. This is because, as I mentioned, IM subtracts 50% from values before showing them in txt:, even though "-profile" creates neutral gray values at 0%.

I'm not sure what happens in test # 6. My result is different to yours. I'll look at this further.

I haven't looked at the CMYK results.

I don't use tificc or Photoshop.

**Reply by Kallys** [reply]
> Your Lab test # 7 is:
> 
> ```
> $ magick "xc:cmyk(100%,0%,0%,0%)" -profile Coated\ FOGRA39\ \(ISO\ 12647-2-2004\).icc -colorspace lab txt:
> ```
> 
> Test # 8 is similar. You are mixing two methods: "-profile" and "-colorspace". Don't do this. The methods are not interchangeable. Unless you take great care, you should not use both methods in the same command.

Yes, I know, this one was a desperate attempt (maybe I was expecting some magick!)...  
In fact, the real problem to me is that IM requires two profiles to perform profile conversion (which makes sense), but as far as I understand, there is no CIELAB profile since it is kind of "absolute" and not device specific.  
So I don't know how to properly convert CMYK with Fogra39 to LAB, and I don't feel really comfortable using a Lab profile found on a forum.

> Tests # 1 to # 4 give similar results. The values seem to be on a scale of -50 to +50. The small differences may come from variations in processing the ICC metadata, or black point compensation, or intents.
> 
> Test # 5 has a* and b* values 50% lower than in # 1 to # 4. This is because, as I mentioned, IM subtracts 50% from values before showing them in txt:, even though "-profile" creates neutral gray values at 0%.

So I need to shift by myself components like this?
```
$ magick "xc:cmyk(100%,0%,0%,0%)" -profile Coated\ FOGRA39\ \(ISO\ 12647-2-2004\).icc -profile GenericLabProfile.icc -channel GB -fx 'u+128/255' +channel txt:
# ImageMagick pixel enumeration: 1,1,65535,cielab
0,0: (37581,22533,-13123.5)  #92CD58054CBC  cielab(57.3449,-39.8229,-51.0642)
```

> I'm not sure what happens in test # 6. My result is different to yours. I'll look at this further.
> 
> I haven't looked at the CMYK results.
> 
> I don't use tificc or Photoshop.

I did not use tificc too.  
But since I was experiencing some issues with those conversions and I learnt that IM uses little CMS 2, I decided to give it a try directly.

**Comment by snibgo** [comment, marked as answer]
> So I don't know how to properly convert CMYK with Fogra39 to LAB, and I don't feel really comfortable using a Lab profile found on a forum.

Personally, I use the profiles made by Elle Stone: https://github.com/ellelstone/elles_icc_profiles/tree/master/profiles

> So I need to shift by myself components like this? ... -fx 'u+128/255'

Something like that. "-fx" is slow, so I suggest "-evaluate". And the exact amount is "(QuantumRange+1)/2". So:
```
-evaluate Add %[fx:(QuantumRange+1)/2]
```

**Comment by Kallys** [comment]
I've updated first message with new attempts for potentially readers.

Profiles distributed by Elle Stone and suggested by @snibgo seems trustful to me.

By now, I'll use following commands to mock up Photoshop behaviour.

**CMYK to CIELAB conversion:**
```$ magick <input> -black-point-compensation -intent relative -profile Coated\ FOGRA39\ \(ISO\ 12647-2-2004\).icc -profile Lab-D50-Identity-elle-V4.icc -channel GB -evaluate Add %[fx:\(QuantumRange+1\)/2] +channel <output>```

**CIELAB to CMYK conversion:**
```$ magick <input> -black-point-compensation -intent relative -profile Lab-D50-Identity-elle-V4.icc -profile Coated\ FOGRA39\ \(ISO\ 12647-2-2004\).icc <output>```

**Reply by Mad4F1** [reply]
Hi. Thanks for this article. However, I am experiencing issues trying to run your last "CMYK to CIELAB" command. Running the following command:
>magick "xc:cmyk(100%,0%,0%,0%)" -black-point-compensation -intent relative -profile Coated_Fogra39L_VIGC_300.icc -profile Lab-D50-Identity-elle-V4.icc -channel GB -evaluate Add %[fx:\(QuantumRange+1\)/2] +channel C:\Users\dino\Pictures\cyan2lab-9.tiff

I'm getting the following errors:
unable to transform colorspace `cmyk(100%,0%,0%,0%)', not an ICC profile, invalid signature (#11) @ warning/profile.c/CMSExceptionHandler/513.
magick: ColorspaceColorProfileMismatch `icc' @ error/profile.c/ProfileImage/1048.
magick: Expected operand at '\(QuantumRange+1\)/2' @ error/fx.c/GetOperand/2416.
magick: invalid argument for option '-evaluate' '' at CLI arg 11 @ error/operation.c/CLISimpleOperatorImage/2352.

My Version: ImageMagick 7.1.1-36 Q16-HDRI x64 852a4e9:20240727

Would really appreciate your help, as this is something that I would really like to be able to do.

**Comment by snibgo** [comment]
> ... not an ICC profile, invalid signature ...

That suggests one of your ICC files is bad.

Your `-channel GB -evaluate Add %[fx:(QuantumRange+1)/2] +channel` works fine for me, IM v7.1.1-20 on Windows 11. Try your command with no `-profile` operations.
